### PR TITLE
Convert sparkplug-payload to typescript

### LINKED ISSUE #138: "TypeError: sparkplugPayload.timestamp.toNumber is not a function"
I am having a problem using the JavaScript Sparkplug Payload, as described on README,
encoded = sparkplug.encodePayload(payload);Normal, but Decoding a payload below.
var decoded = sparkplug.decodePayload(encoded); Start an error：

chunk-vendors.js:3874 [Vue warn]: Error in v-on handler: "TypeError: sparkplugPayload.timestamp.toNumber is not a function"

found in

---> <Root>

[system] TypeError: sparkplugPayload.timestamp.toNumber is not a function
    at Object.push.D:\MyTestStorage\Vuejs\HBuilderProjects\sparkplug-client-test\node_modules\sparkplug-payload\lib\sparkplugbpayload.js.exports.decodePayload (pages-index-index.js:10486)
    at VueComponent.test (VM441 index.js!D:\MyTestStorage\Vuejs\HBuilderProjects\sparkplug-client-test\pages\index\index.vue:37)
    at click (pages-index-index.js:63)
    at invokeWithErrorHandling (chunk-vendors.js:5117)
    at invoker (chunk-vendors.js:5442)
    at invokeWithErrorHandling (chunk-vendors.js:5117)
    at HTMLElement.invoker (chunk-vendors.js:5438)
    at HTMLElement.original._wrapper (chunk-vendors.js:10320)

sparkplugbpayload.js

image

My code:
`

                            test() {

				var sparkplug = require('sparkplug-payload').get("spBv1.0"),
					payload = {
						"timestamp": new Date().getTime(),
						"metrics": [{
							"name": "intMetric",
							"value": 1,
							"type": "Int32"
						}]
					},
					encoded = sparkplug.encodePayload(payload);

				console.log(encoded);
				var decoded = sparkplug.decodePayload(encoded);
				console.log(decoded);
			}
`

### PR DESCRIPTION
**Pull request by ferm10n**
- fixes #138 (there's no `toNumber()` function in javascript, I just replaced it with a `Number(...)` wrap)
- I tried to break up the changes incrementally as much as I could to preserve git history
- sparkplugpayload.js (.ts now) does not get wrapped in an anonymous function. I assume that's ok since it was originally using commonjs (and still builds to commonjs)
- extract protobuf definition into its own file.
  - utilize protobuf CLI to auto generate types and static code
-  judging from the readme example, it looked like the user was supposed to be able to just set value and type and the lib would figure out the rest. This means there's a need for intermediate types between those given by the user and those generated by the .proto file. these are the `U___` types
    - > *This in particular needs attention, since the README does not specify example usage for anything except simple metrics*

### CODE REVIEW

**Comment by gaillarddamien** [comment]
Hi,
I'd like to point out that `toNumber()` comes from the [Long](https://www.npmjs.com/package/long) library.
`timestamp` is a `uint64` in protobuf, and might not fit as a JS `Number`: therefore it is decoded as a `Long` object.

Whether you call `timestamp.toNumber()` or `Number(timestamp)` will do fine here:
- we are unlikely to use timestamps bigger than [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER).
- it has a `toString()` method too: you are actually calling `Number(timestamp.toString())` here, AFAIK. (you'd get `NaN` otherwise)

BTW I see other 64bit-long fields such as `seq`, `alias`, `numOfColumns` which are either converted to `Number`, or left as `Long` objects. I guess it is assumed `seq` is never going to be bigger than `MAX_SAFE_INTEGER`?

**Comment by ferm10n** [comment]
To be honest, I don't really understand why toNumber was called in the first place instead of doing a direct assignment.
take this for example

https://github.com/eclipse/tahu/blob/72a9763841396053190a61f4f178ad9b93d4072e/javascript/core/sparkplug-payload/lib/sparkplugbpayload.js#L667-L669

`metric.timestamp` and `protoMetric.timestamp` are both typed as `number | Long.Long`... so it could also be done like this right?
```ts
if (protoMetric.hasOwnProperty("timestamp")) {
    metric.timestamp = protoMetric.timestamp;
}
```

File: javascript/core/sparkplug-payload/lib/SparkplugPayload.proto [hunk 1]
```diff
@@ -0,0 +1,125 @@
+package org.eclipse.tahu.protobuf;
```
Comment (wes-johnson) [inline javascript/core/sparkplug-payload/lib/SparkplugPayload.proto:1]: Does this file really need to be duplicated here?

**Review by wes-johnson — CHANGES_REQUESTED** [review]
I just have the one question on the duplicated .proto file.

File: javascript/core/sparkplug-payload/lib/SparkplugPayload.proto (hunk 1 above)
Comment (ferm10n) [reply javascript/core/sparkplug-payload/lib/SparkplugPayload.proto:1]: ah! didn't realize there was an existing proto file.

**Comment by ferm10n** [comment]
I also added declaration files to be published with the package

**Review by wes-johnson — APPROVED** [review]

**Pull request merged**